### PR TITLE
Support new extension points for ARM resources

### PR DESCRIPTION
**Pull request by matthchr**
- Hook for deletion
- Hook for after-successful-creation
- Hook for claim

### CODE REVIEW

File: v2/pkg/genruntime/extensions/claimer.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+// Claimer can be implemented to customize how the reconciler claims a resource
```
Comment (theunrepentantgeek) [inline v2/pkg/genruntime/extensions/claimer.go:17]: Do you need to doc what Claiming is about?

File: v2/pkg/genruntime/extensions/successful_resource_modifier.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+// SuccessfulResourceModifier can be implemented to customize the resource upon successful creation
+type SuccessfulResourceModifier interface {
+	// Success modifies the resource based on a successful creation
+	Success(obj genruntime.ARMMetaObject) error
+}
```
Comment (theunrepentantgeek) [inline v2/pkg/genruntime/extensions/successful_resource_modifier.go:20]: What's the scenario where this extension point is needed?

**Review by theunrepentantgeek — APPROVED** [review]

File: v2/pkg/genruntime/extensions/claimer.go (hunk 1 above)
Comment (matthchr) [reply v2/pkg/genruntime/extensions/claimer.go:17]: Its part of the `Reconciler` interface we define and is described in that interface already:
```
	// Claim performs resource specific claim actions. This must be idempotent.
	// A standard finalizer is added to all resources, Claim
	// should deal with any resource specific claiming actions (such as setting a resource ID annotation, etc).
	// If Claim returns an error then reconciliation will be retried according to the returned Ready condition until
	// no error is returned. Once Claim succeeds CreateOrUpdate is called.
	Claim(
		ctx context.Context,
		log logr.Logger,
		eventRecorder record.EventRecorder,
		obj MetaObject) error
```

Do you think we need more than that?

File: v2/internal/reconcilers/arm/deleter.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func deleteResource(
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/reconcilers/arm/deleter.go:19]: Ive moved this back into `azure_generic_arm_reconciler` (which is where it is used, so that should be more clear), and added the following comment:
```
// deleteResource deletes a resource in ARM. This function is used as the default deletion handler and can
// have its behavior modified by resources implementing the genruntime.Deleter extension
```

File: v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go [hunk 4]
```diff
@@ -356,6 +329,12 @@ func (r *azureDeploymentReconcilerInstance) handleCreatePollerSuccess(ctx contex
 		return ctrl.Result{}, err
 	}
 
+	onSuccess := extensions.CreateSuccessfulResourceModifier(r.Extension, r.Log)
```
[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go:332]: Yeah I don't like the name either. The hook is: "When a resource has been created successfully". We could call it something more action-oriented like `OnSuccessfulResourceCreation`. The issue is that's not really `CreateX` which is the pattern we have for other extension handlers, where X is the name of the extension.

`CreateResourceCreationHandler`?

[the remark this answers is not in the document]
Comment (matthchr) [reply v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go:332]: I've moved this to `SuccessfulCreationHandler` + `CreateSuccessfulCreationHandler`

File: v2/pkg/genruntime/extensions/successful_resource_modifier.go (hunk 2 above)
Comment (matthchr) [reply v2/pkg/genruntime/extensions/successful_resource_modifier.go]: Because `subscriptionAliases` don't have the right url to assign child resources (like RoleAssignment) to, I've got in another branch the addition of a concept called `ChildResourceIDOverride`, which is an annotation that can be added to an object. If that annotation is added to the object, when child resources are attached to that object, instead of using the normal ID/Type of the object they just use the URL in the annotation. This hook is so that I can manually add that annotation to the `subscriptionAlias` type, which then causes it to be able to correctly have child resources created off of the subscription (rather than off of the alias).

Basically it's a way to combine two separate resources into 1. It's somewhat hacky so we shouldn't really use it regularly, but it saves us from having to implement our own reconciler for subscription, it can just use the normal reconciler.

File: v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go]: That works. :+1:

File: v2/pkg/genruntime/extensions/claimer.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/pkg/genruntime/extensions/claimer.go:17]: No, that'll do. I missed that.

File: v2/pkg/genruntime/extensions/successful_resource_modifier.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/pkg/genruntime/extensions/successful_resource_modifier.go]: Good use of an extension point.
